iocell/src/main/scala/chisel/IOCell.scala

```diff
@@ -0,0 +1,188 @@
+  val io: DigitalOutIOCellBundle
+}
+
+class ExampleAnalogIOCell extends AnalogIOCell {
+  val io = IO(new AnalogIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalGPIOCell extends DigitalGPIOCell {
+  val io = IO(new DigitalGPIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalInIOCell extends DigitalInIOCell {
+  val io = IO(new DigitalInIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalOutIOCell extends DigitalOutIOCell {
+  val io = IO(new DigitalOutIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+object IOCell {
+
+  def exampleAnalog() = Module(new ExampleAnalogIOCell)
+  def exampleGPIO() = Module(new ExampleDigitalGPIOCell)
+  def exampleInput() = Module(new ExampleDigitalInIOCell)
+  def exampleOutput() = Module(new ExampleDigitalOutIOCell)
+
+  def generateIOFromSignal[T <: Data](coreSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): (T, Seq[IOCell]) =
+  {
+    val padSignal = IO(DataMirror.internal.chiselTypeClone[T](coreSignal))
+    val iocells = IOCell.generateFromSignal(coreSignal, padSignal, name, inFn, outFn, anaFn)
+    (padSignal, iocells)
+  }
+
+  def generateFromSignal[T <: Data](coreSignal: T, padSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): Seq[IOCell] =
+  {
+    coreSignal match {
+      case coreSignal: Analog => {
+        if (coreSignal.getWidth == 0) {
+          Seq()
+        } else {
+          require(coreSignal.getWidth == 1, "Analogs wider than 1 bit are not supported because we can't bit-select Analogs (https://github.com/freechipsproject/chisel3/issues/536)")
+          val iocell = anaFn()
+          name.foreach(n => iocell.suggestName(n))
+          iocell.io.core <> coreSignal
+          padSignal <> iocell.io.pad
+          Seq(iocell)
+        }
+      }
+      case coreSignal: Clock => {
+        DataMirror.directionOf(coreSignal) match {
+          case ActualDirection.Input => {
+            val iocell = inFn()
+            name.foreach(n => iocell.suggestName(n))
+            coreSignal := iocell.io.i.asClock
+            iocell.io.ie := true.B
+            iocell.io.pad := padSignal.asUInt.asBool
+            Seq(iocell)
+          }
+          case ActualDirection.Output => {
+            val iocell = outFn()
+            name.foreach(n => iocell.suggestName(n))
+            iocell.io.o := coreSignal.asUInt.asBool
+            iocell.io.oe := true.B
+            padSignal := iocell.io.pad.asClock
+            Seq(iocell)
+          }
+          case _ => throw new Exception("Unknown direction")
```

Comment:
I think this exception should be more descriptive.
"Clock IO does not have direction and cannot be matched to an IOCell"
or something like that.

Reply:
Sure, will fix.